infrastructure/smart-contracts/contracts/upgrades/referee/Referee10.sol

```diff
@@ -759,31 +761,35 @@ contract Referee10 is Initializable, AccessControlEnumerableUpgradeable {
         emit UnstakeV1(msg.sender, amount, stakedAmounts[msg.sender]);
     }
 
-    function stakeKeys(address pool, address staker, uint256[] memory keyIds, bool isAdminStake) external onlyPoolFactory {
+    function stakeKeys(address pool, address staker, uint256 keyAmount, bool isAdminStake) external onlyPoolFactory {
         require(isAdminStake || stakingEnabled, "52");
 
-        uint256 keysLength = keyIds.length;
         // Check if the pool has enough capacity to stake the keys
-        require(assignedKeysToPoolCount[pool] + keysLength <= maxKeysPerPool, "43");
-
-        uint256 currentChallenge = challengeCounter - 1;
+        require(assignedKeysToPoolCount[pool] + keyAmount <= maxKeysPerPool, "43");
 
         NodeLicense nodeLicenseContract = NodeLicense(nodeLicenseAddress);
+        uint256 ownedKeyAmount = nodeLicenseContract.balanceOf(staker);
+
+        require(ownedKeyAmount >= assignedKeysOfUserCount[staker] + keyAmount, "45");
+
+        // DEPRECATED We do not track keyIds anymore, we only check if the staker has enough unstaked keys available
         // Assign the keys to the pool
-        for (uint256 i = 0; i < keysLength; i++) {
-            uint256 keyId = keyIds[i];
-            // Check if the key is not already assigned to a pool
-            require(assignedKeyToPool[keyId] == address(0), "44");
-            if(!isAdminStake){
-                // If not admin stake, check if the staker is the owner of the key
-                require(nodeLicenseContract.ownerOf(keyId) == staker, "45");
-            }
-            assignedKeyToPool[keyId] = pool;
-        }
+        // for (uint256 i = 0; i < keysLength; i++) {
+        //     uint256 keyId = keyIds[i];
+        //     // Check if the key is not already assigned to a pool
+        //     require(assignedKeyToPool[keyId] == address(0), "44");
+        //     if(!isAdminStake){
+        //         // If not admin stake, check if the staker is the owner of the key
+        //         require(nodeLicenseContract.ownerOf(keyId) == staker, "45");
+        //     }
+        //     assignedKeyToPool[keyId] = pool;
+        // }
 
         // Update the user and pool staked key counts
-        assignedKeysToPoolCount[pool] += keysLength;
-        assignedKeysOfUserCount[staker] += keysLength;
+        assignedKeysToPoolCount[pool] += keyAmount;
+        assignedKeysOfUserCount[staker] += keyAmount;
+
+        uint256 currentChallenge = challengeCounter - 1;
```

Comment:
Should this check if we're on challenge 0 to avoid underflow in test suite?

Reply:
I think we are submittting a challenge in the test suite to represent the flow of state getting added on production.
The testcases all passed in the other PR.